### PR TITLE
New endpoints to get equipments ids

### PR DESCRIPTION
**Pull request by anistouri**

### CODE REVIEW

**Comment by SlimaneAmar** [comment]
I think it's better to have a generic path like this :
/networks/{networkUuid}/variants/{variantId}/equipmentsIds?type=''

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java [hunk 1]
```diff
@@ -768,6 +778,20 @@ public List<LineMapData> getLines(UUID networkUuid, String variantId, List<Strin
         }
     }
 
+    public List<String> getLinesIds(UUID networkUuid, String variantId, List<String> substationsIds) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/network/map/NetworkMapService.java:781]: Use this generic form
```
    public List<String> getEquipmentsIds(UUID networkUuid, String variantId, List<String> substationsIds, Class<? extends Connectable> equipmentClass) {
        Network network = getNetwork(networkUuid, getPreloadingStrategy(substationsIds), variantId);
        if (substationsIds == null) {
            return network.getConnectableStream(equipmentClass)
                .map(Connectable::getId)
                .collect(Collectors.toList());
        } else {
            return substationsIds.stream()
                .flatMap(substationId -> network.getSubstation(substationId).getVoltageLevelStream())
                .flatMap(voltageLevel -> voltageLevel.getConnectableStream(equipmentClass))
                .map(Connectable::getId)
                .collect(Collectors.toList());
        }
    }
```

File: src/main/java/org/gridsuite/network/map/NetworkMapController.java [hunk 2]
```diff
@@ -81,6 +81,51 @@ public class NetworkMapController {
...
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "Equipments ids")})
+    public @ResponseBody List<String> getEquipmentsIds(@Parameter(description = "Network UUID") @PathVariable("networkUuid") UUID networkUuid,
+                                                       @Parameter(description = "Variant Id") @RequestParam(name = "variantId", required = false) String variantId,
+                                                       @Parameter(description = "Substations id") @RequestParam(name = "substationId", required = false) List<String> substationsIds,
+                                                       @Parameter(description = "equipment type") @RequestParam(name = "equipmentType", required = true) MapConstant.EquipmentType equipmentType) {
+        return getIds(networkUuid, variantId, substationsIds, equipmentType);
+
+    }
+
+    private List<String> getIds(UUID networkUuid, String variantId, List<String> substationsIds, MapConstant.EquipmentType equipmentType) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/network/map/NetworkMapController.java:95]: Rename by `getEquipmentsIds` et move it the class service

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/test/resources/voltage-level-VLGEN-equipments.json [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    "id": "DL1",
+    "name": "DL1",
+    "type": "DANGLING_LINE"
+  },
+  {
+    "id": "LINE3",
+    "name": null,
+    "type": "LINE"
+  }
+]
```
Comment (SlimaneAmar) [inline src/test/resources/voltage-level-VLGEN-equipments.json:52]: Add new line

File: src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java [hunk 4]
```diff
@@ -719,17 +720,25 @@ private String resourceToString(String resource) throws IOException {
         return new String(ByteStreams.toByteArray(getClass().getResourceAsStream(resource)), StandardCharsets.UTF_8);
     }
 
-    private String buildUrlForList(String equipments, String variantId, List<String> immutableListSubstationIds) {
+    private String buildUrlForList(String equipment, String variantId, List<String> immutableListSubstationIds, boolean onlyIds) {
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java:723]: `equipmentTypeInfo` in place of `equipment` ?

File: src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java [hunk 5]
```diff
@@ -719,17 +720,25 @@ private String resourceToString(String resource) throws IOException {
...
         List<String> substationsIds = immutableListSubstationIds == null ? List.of() : immutableListSubstationIds.stream().collect(Collectors.toList());
-        StringBuffer url = new StringBuffer("/v1/networks/{networkUuid}/" + equipments);
+        //StringBuffer url = new StringBuffer("/v1/networks/{networkUuid}/" + equipments + (onlyIds ? "/ids" : ""));
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java:725]: ?

File: src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java [hunk 6]
```diff
@@ -782,30 +791,38 @@ private void failingTestForElement(String equipments, UUID networkUuid, String v
                 .andExpect(status().isNotFound());
     }
 
-    private String buildUrlVoltageLevels(String variantId, List<String> immutableListSubstationIds) {
+    private String buildUrlVoltageLevels(String variantId, List<String> immutableListSubstationIds, boolean onlyIds) {
         List<String> substationsIds = immutableListSubstationIds == null ? List.of() : immutableListSubstationIds.stream().collect(Collectors.toList());
-        StringBuffer url = new StringBuffer("/v1/networks/{networkUuid}/voltage-levels");
+        //StringBuffer url = new StringBuffer("/v1/networks/{networkUuid}/voltage-levels" + (onlyids ? "/ids" : ""));
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java:796]: ?

File: src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java [hunk 7]
```diff
@@ -1307,61 +1498,68 @@ public void shouldReturnAnErrorInsteadOfBusbarSectionMapData() throws Exception
...
     public void shouldReturnVoltageLevelsAndEquipments() throws Exception {
-        succeedingTestForList("voltage-levels-equipments", NETWORK_UUID, null, null, resourceToString("/voltage-levels-equipments-map-data.json"));
-        succeedingTestForList("voltage-levels-equipments", NETWORK_UUID, VARIANT_ID, null, resourceToString("/voltage-levels-equipments-map-data-in-variant.json"));
+        succeedingTestForList("voltage-levels-equipments", NETWORK_UUID, null, null, false, resourceToString("/voltage-levels-equipments-map-data.json"));
+        succeedingTestForList("voltage-levels-equipments", NETWORK_UUID, VARIANT_ID, null, false, resourceToString("/voltage-levels-equipments-map-data-in-variant.json"));
+    }
+
+    @Test
+    public void shouldReturnVoltageLevelEquipments() throws Exception {
+
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java:1507]: Remove blank line

File: src/main/java/org/gridsuite/network/map/utils/EquipmentType.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package org.gridsuite.network.map.utils;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/network/map/utils/EquipmentType.java:1]: public enum EquipmentType {
    NETWORK(IdentifiableType.NETWORK.name(), Network.class),
    SUBSTATION(IdentifiableType.NETWORK.name(), Substation.class),
    VOLTAGE_LEVEL(IdentifiableType.VOLTAGE_LEVEL.name(), VoltageLevel.class),
    HVDC_LINE(IdentifiableType.HVDC_LINE.name(), HvdcLine.class),
    BUSBAR_SECTION(IdentifiableType.BUSBAR_SECTION.name(), BusbarSection.class),
    LINE(IdentifiableType.LINE.name(), Line.class),
    TWO_WINDINGS_TRANSFORMER(IdentifiableType.TWO_WINDINGS_TRANSFORMER.name(), TwoWindingsTransformer.class),
    THREE_WINDINGS_TRANSFORMER(IdentifiableType.THREE_WINDINGS_TRANSFORMER.name(), ThreeWindingsTransformer.class),
    GENERATOR(IdentifiableType.GENERATOR.name(), Generator.class),
    BATTERY(IdentifiableType.BATTERY.name(), Battery.class),
    LOAD(IdentifiableType.LOAD.name(), Load.class),
    SHUNT_COMPENSATOR(IdentifiableType.SHUNT_COMPENSATOR.name(), ShuntCompensator.class),
    DANGLING_LINE(IdentifiableType.DANGLING_LINE.name(), DanglingLine.class),
    STATIC_VAR_COMPENSATOR(IdentifiableType.STATIC_VAR_COMPENSATOR.name(), StaticVarCompensator.class),
    LCC_CONVERTER_STATION("LCC_CONVERTER_STATION", LccConverterStation.class),
    VSC_CONVERTER_STATION("VSC_CONVERTER_STATION", VscConverterStation.class);

    private String typeName;

    private Class typeClass;

    public Class getTypeClass() {
        return typeClass;
    }

    public String getName() {
        return typeName;
    }

    private EquipmentType(String typeName, Class typeClass) {
        this.typeName = typeName;
        this.typeClass = typeClass;
    }
}

File: src/main/java/org/gridsuite/network/map/NetworkMapController.java [hunk 9]
```diff
@@ -261,6 +282,16 @@ public class NetworkMapController {
         return networkMapService.getVoltageLevelsAndConnectable(networkUuid, variantId, substationsIds);
     }
 
+    @GetMapping(value = "/networks/{networkUuid}/voltage-level-equipments/{voltageLevelId}", produces = APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (anistouri) [reply src/main/java/org/gridsuite/network/map/NetworkMapController.java:285]: it's to get the equipments of a voltage level. we don't need to collect all the equipments we don't need

File: src/test/resources/voltage-level-VLGEN-equipments.json (hunk 3 above)
Comment (anistouri) [reply src/test/resources/voltage-level-VLGEN-equipments.json:52]: This is the answer we are waiting for

File: src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java (hunk 4 above)
Comment (anistouri) [reply src/test/java/org/gridsuite/network/map/NetworkMapControllerTest.java:723]: I named it equipmentType to be consistent with the others

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java [hunk 10]
```diff
@@ -777,6 +788,22 @@ public List<LineMapData> getLines(UUID networkUuid, String variantId, List<Strin
         }
     }
 
+    public List<String> getEquipmentsIds(UUID networkUuid, String variantId, List<String> substationsIds, Class<? extends Connectable> equipmentClass) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/network/map/NetworkMapService.java:791]: Private

File: src/main/java/org/gridsuite/network/map/model/EquipmentType.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
+package org.gridsuite.network.map.model;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/network/map/model/EquipmentType.java:1]: Copyright

Author

File: src/main/java/org/gridsuite/network/map/model/EquipmentType.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.powsybl.iidm.network.*;
+
+public enum EquipmentType {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/network/map/model/EquipmentType.java:5]: public enum EquipmentType {
    NETWORK(IdentifiableType.NETWORK, Network.class),
    SUBSTATION(IdentifiableType.NETWORK, Substation.class),
    VOLTAGE_LEVEL(IdentifiableType.VOLTAGE_LEVEL, VoltageLevel.class),
    HVDC_LINE(IdentifiableType.HVDC_LINE, HvdcLine.class),
    BUSBAR_SECTION(IdentifiableType.BUSBAR_SECTION, BusbarSection.class),
    LINE(IdentifiableType.LINE, Line.class),
    TWO_WINDINGS_TRANSFORMER(IdentifiableType.TWO_WINDINGS_TRANSFORMER, TwoWindingsTransformer.class),
    THREE_WINDINGS_TRANSFORMER(IdentifiableType.THREE_WINDINGS_TRANSFORMER, ThreeWindingsTransformer.class),
    GENERATOR(IdentifiableType.GENERATOR, Generator.class),
    BATTERY(IdentifiableType.BATTERY, Battery.class),
    LOAD(IdentifiableType.LOAD, Load.class),
    SHUNT_COMPENSATOR(IdentifiableType.SHUNT_COMPENSATOR, ShuntCompensator.class),
    DANGLING_LINE(IdentifiableType.DANGLING_LINE, DanglingLine.class),
    STATIC_VAR_COMPENSATOR(IdentifiableType.STATIC_VAR_COMPENSATOR, StaticVarCompensator.class),
    LCC_CONVERTER_STATION(IdentifiableType.HVDC_CONVERTER_STATION, LccConverterStation.class),
    VSC_CONVERTER_STATION(IdentifiableType.HVDC_CONVERTER_STATION, VscConverterStation.class);

    private final IdentifiableType identifiableType;

    private final Class<? extends Identifiable> typeClass;

    public IdentifiableType getIdentifiableType() {
        return identifiableType;
    }
    
    public Class<? extends Identifiable> getTypeClass() {
        return typeClass;
    }

    EquipmentType(IdentifiableType identifiableType, Class<? extends Identifiable> typeClass) {
        this.identifiableType = identifiableType;
        this.typeClass = typeClass;
    }
}

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java [hunk 13]
```diff
@@ -1140,7 +1202,48 @@ public List<LineMapData> getMapLines(UUID networkUuid, String variantId, List<St
...
+        Network network = getNetwork(networkUuid, getPreloadingStrategy(substationsIds), variantId);
+        if (substationsIds == null) {
+            return network.getHvdcLineStream()
+                    .map(HvdcLine::getId).collect(Collectors.toList());
+        } else {
+            return substationsIds.stream()
+                    .flatMap(substationsId -> network.getSubstation(substationsId).getVoltageLevelStream())
+                    .flatMap(voltageLevel -> voltageLevel.getConnectableStream(HvdcConverterStation.class))
+                    .filter(hvdcConverterStation -> hvdcConverterStation.getHvdcLine() != null)
+                    .map(HvdcConverterStation::getHvdcLine)
```
Comment (YenguiSeddik) [inline src/main/java/org/gridsuite/network/map/NetworkMapService.java:1218]: you can filter after getting the hvdcLine
.map(HvdcConverterStation::getHvdcLine)
.filter(Objects::nonNull)

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java [hunk 14]
```diff
@@ -1079,32 +1090,77 @@ public List<VscConverterStationMapData> getVscConverterStations(UUID networkUuid
...
 
+    public List<HvdcLineMapData> getHvdcLines(UUID networkUuid, String variantId, List<String> substationsId) {
+        Network network = getNetwork(networkUuid, getPreloadingStrategy(substationsId), variantId);
+        if (substationsId == null) {
+            return network.getHvdcLineStream()
+                    .map(NetworkMapService::toMapData).collect(Collectors.toList());
+        } else {
+            Set<HvdcLineMapData> res = new LinkedHashSet<>();
+            substationsId.stream().forEach(id ->
+                    network.getSubstation(id).getVoltageLevelStream().forEach(v ->
```
Comment (YenguiSeddik) [inline src/main/java/org/gridsuite/network/map/NetworkMapService.java:1101]: you can do forEach directly without stream

Comment (anistouri) [reply src/main/java/org/gridsuite/network/map/NetworkMapService.java:1101]: Indeed, I didn't pay attention, it was in the old code. I'll correct it.

**Review by SlimaneAmar — APPROVED** [review]

**Review by YenguiSeddik — APPROVED** [review]

**Pull request merged**